rfc/0003-tests-arch.md

```diff
@@ -0,0 +1,180 @@
+Kubernetes operator will have the expected features working.
+
+These tests are trigger when we have new releases (or release candidate) of
+any of the Kubewarden stack.  It should follow some steps:
+
+1. Setup Kubernetes cluster
+  Kubewarden is expected to run in multiple Kubernetes versions and machine
+  architectures. So, the end-to-end tests should ensure that all the supported
+  versions and architectures are tested.
+2. Install Kuberwarden stack on the cluster from step 1
+  The version to be tested should be installed. This means that when a component
+  is released. It should be tested against all the other components versions which
+  it supposed to work together.
+3. Run basic end-to-end tests
+  Validate if the basic features are working as expected
+4. Run reconfiguration tests
+  Validates if the user is able to reconfigure a Kubewarden stack installed
+5. Run upgrade tests
+  Validate if the user is able to upgrade and continue to work with the latest release.
+
+Steps 3,4 and 5 may run in parallel.
+
+Considering the tests will validate multiple components of the Kubewarden stack
+together, it desirable to keep all the end-to-end test files in a separated
+repository. This repository should contains all the yaml files, scripts, and
+helper code to run the tests. This also helps to avoid duplicating tests that
+validate some feature that impacts multiple components.
+
+The end-to-end tests repository should be independent. Which means that it should
+run the tests independently of the tests and development cycle of the other
+Kubewarden repositories. But every time a new component release is available
+for testing, the end-to-end tests should be trigger automatically with no human
+intervention.  Besides that, running the end-to-end test on demand and in pre defined
+periods is also desirable.
+
+We could also configure the end-to-end tests to run against the latest version of
+all Kubewarden components. Thus, spotting some possible issues before a official
+release candidate.
+
+Considering the high workload nature of the end-to-end tests it may be necessary
+run them in a dedicated runner outside from Github hosted runners. From past
+experiences, the runners from Github could be slow and the tests failed due
+timeout issues.
+
+This tests can be written using bash and frameworks similar to [bats](https://github.com/bats-core/bats-core/).
+
+The developers should be able to run the same end-to-end tests locally with simple
+commands like:
+
+```bash
+# in the end-to-end test respository directory
+make test
+```
+
+## Policy tests
+
+It's also desirable to have tests to validate policies behaviour. So, it is
+necessary provide policy authors some tools to help them testing their policies.
+There is an initial [tool](https://github.com/kubewarden/setup-kubewarden-cluster-action)
+for that. It's a Github action which creates a K3D cluster and installs Kubewarden
+on it.
```

Comment:
I would like to experiment a little with a well-known structure for test cases/expected responses in policies trees.

If we settle on a common tree structure for tests and expected results on policies, it would be straightforward to implement something along the lines of `kwctl test`, where one can run inside a policy folder, and get all those e2e tests *for the policy* executed by kwctl.

I'm derailing from the main goal of this RFC though.

Reply:
I'm not sure if forcing the user to follow a tree structure to run tests is a good idea. It does not seem flexible to me. Furthermore, what's the advantage of this against the whole universe of tests frameworks with a lot more features that they can use?

Reply:
The idea would be that policies can have a well known folder structure that could have test cases in an automated way. The main advantage of having this structure on all policies would be that besides being e2e tests they also serve as documentation for policy users.

Also, we could integrate the Kubewarden Hub so it can link directly to the test directory, and even in the future show them in a different way in the UI, so they serve as the real documentation of the policy. I think is superior than README.md based documentation that can get easily out of date. e2e tests serve both for validating the policy does what I expect, and to let me know how the policy works.

Reply:
If a certain policy wants to ignore this structure and perform its own way of e2e testing it's perfectly fine, only that the integration with the UI and "e2e tests as policy documentation for free" would not work out of the box for that policy.

It helps to have a more consistent story around policies. No matter their language, framework...

Reply:
I agree in coming up with a well-known structure for test cases/expected responses. In fact, I would like that structure to be applicable to all parts of the project, not only kwctl/policies.

I think the biggest problem with a production CI is coming up with an interface for it between the CI logic (local debugger) and all the different projects that form Kubewarden. The rest (whether we use bats, a different test framework, or several depending on the task) should be interchangeable.

The local debugger interfaces with each of the tests or each kubewarden component (policy-server, kwctl, controller, policies..) in a specific way. And the local debugger loop needs to keep working, no matter if it is asked to run against a tuple of old versions of components, a tuple of the latest components, or a mix. This protects us about needing to backport CI testcases, or do weird things in the future, once we have several versions released and receiving support. It protects us from needing to redo parts of the Github Actions workflows if we want to 

IMHO, this is the most difficult question to answer, as it freezes the interface, and makes us commit to an architecture. These kind of questions are the important ones. Other kind of questions, where one can iterate/backtrack or substitute things in the architecture without changing its interface, are "easy" questions.

I would like to answer the following questions:

- How do I known what tuple of Kubewarden components (kwctl, policy-server, controller, policies, etc) is currently under test? Which commit hash? how to add more to the list, or substitute one?
- Where are the testcases for tests related to only 1 component? (I suppose in the repo of the component under test)
- Where are the testcases for tests that exercise several components (maybe in a repo of a component that depends on both? in a different repo?)
- How do I add tests so they can be reviewed together with the new feature that enables it? Easy if tests are in the repo of the component, just add them to the PR. What happens in other cases?
- How does the debugger engine loop (a simple logic loop) go through all the components and triggers their testscases? How does it call them? How does it get info back that they passed? How does it print errors? How does it skip some if needed?
- How can we make sure that I can refactor the tests of 1 component to use a different test framework (from bats to ginkgo, for example), without needing to change anything besides that component repo?
- How can I use `git bisect` between different tuples of components?

Reply:
```
I agree in coming up with a well-known structure for test cases/expected responses. In fact, I would like that structure to be applicable to all parts of the project, not only kwctl/policies.

I think the biggest problem, from my past experiences, with a production CI is coming up with an interface for it between the CI logic (local debugger) and all the different projects that form Kubewarden. The rest of questions don't freeze the interface (whether we use bats, a different test framework, or several depending on the task).

The local debugger interfaces in a specific way with each of the tests or each kubewarden component (policy-server, kwctl, controller, policies..) that it depends on. The local debugger loop needs to keep working, no matter if it is asked to run against a tuple of old versions of components, a tuple of the latest components, or a mix. This protects us about needing to backport testcases, or do weird things in the future, once we have several versions released and receiving support.

The local debugger also provides an external interface, so one can run it locally or via a scheduler (GH Actions). This interface needs to be frozen too, as it protects us from needing to redo parts of the Github Actions workflows on components or backporting changes to the debugger, on old branches of components already released.

IMHO, the most difficult questions to answer, are the ones that freeze the interfaces of the local debugger:

-   the internal one to each component and its testcases
-   the external one that gets used by schedulers (git bisect, GH Actions..)  
    Those make us commit to an architecture. These kind of questions are the important ones, and the other ones are "easy" questions, as we don't have the need to get it right early, and we can iterate on the solutions.
```

I understand the concerns and I agree with you (afaics, this is a common issue in the industry. I find  a similar complain about that in an article from this [repo](https://github.com/asatarin/testing-distributed-systems)). But I think we should keep something in mind, the tests will require some maintenance. I do not think that they will work forever without some care. Trying to minimize this overhead we can start prioritizing  unit tests  and component level . 

Let's add more complex E2E testing in the future after we mature how to approach this using this script locally during development (Does it make sense?). To reflect that, I can update the E2E section to express this or simple remove it completely.  Otherwise, if we start without a clear way to do things, it will be a excuse to not adding tests or ignoring them if they fail.

```
I would like to answer the following questions:

-   How do I known what tuple of Kubewarden components (kwctl, policy-server, controller, policies, etc) is currently under test? Which commit hash? how to add more to the list, or substitute one?
```
I do not expect you define the commit hash version for the E2E tests. These tests should simulate the user usage. So, they will not allow the developer to define this level of granularity. What you should be able to do is to define the container images used to deploy the Kubewarden stack (e.g. changing the helm chart configuration). However, you can try individual commits if you publish a container image for that commit somewhere and change the helm chart value to use that container image.
```
-   How do I known what tuple of Kubewarden components (kwctl, policy-server, controller, policies, etc) is currently under test? Which commit hash? how to add more to the list, or substitute one?
```
 As E2E simulates the user, it's not expected to allow specific the commit version for all the components. You will be allowed to specific the container image as we do in the helm charts and yaml files.  We can even change the script to configure this values on the fly something like:
 ```bash
 POLICY_SERVER_VERSION=v0.2.5 make runtests
 # or
 POLICY_SERVER_IMAGE=ghrc.io/user/policy-server:afqwgrnad make runtests
 ```

```
-   Where are the testcases for tests related to only 1 component? (I suppose in the repo of the component under test)
```
Yes. If the test is exclusively for one component it's not necessary to add it in the E2E tests repo. Of course, You can add tests in the E2E to add an extra level of testing. But this will not test the fix at "low" level. You will add the test to validate the behavior of updated component together with the whole stack.  
```
-   Where are the testcases for tests that exercise several components (maybe in a repo of a component that depends on both? in a different repo?)
```
If you want to exercise how different components work together,  add some E2E tests for the feature in the E2E tests repository. I do not thing this is necessary for all the features. But this exercise will test how the components behave together. It will NOT validate how the communication between them happens.  
```
-   How does the debugger engine loop (a simple logic loop) go through all the components and triggers their testscases? How does it call them? How does it get info back that they passed? How does it print errors? How does it skip some if needed
```
That will depend on what we will use to develop tests. Currently, we are using `bats` which already handle all of these for us. If we decided to go with pure shell, we need to implement that by ourselves. If we decided to use different libraries, we can use what their provide.
```
-   How can we make sure that I can refactor the tests of 1 component to use a different test framework (from bats to ginkgo, for example), without needing to change anything besides that component repo?
```
For the component repo, as long as the developer abstracts the two different test frameworks under a helper script or Makefile target, I do not see that as a problem. If the first test suite fails, skip the next one. I do not see the need of unify the test results outputs for now. If the tests passed, great. If it fails, the execution will stop before start the next test suite written in another test framework. 
```
-   How can I use `git bisect` between different tuples of components?
```
For the component test (those one living in the component repo), It should work out of the box.

For the E2E test, it uses versions available somewhere. So, if you are running a bisect,  you can publish the version under the bisect somewhere (probably a registry) and refer it in the E2E tests scripts as described in your first two questions.

Reply:
> Let's add more complex E2E testing in the future after we mature how to approach this using this script locally during development (Does it make sense?). To reflect that, I can update the E2E section to express this or simple remove it completely. 

I'm fine with that. I suppose we would need a follow-up RFC for e2e-tests in the future. But I would like to point that the e2e-tests are the only ones that must have a defined backwards-compatible interface before being on production.

> Yes. If the test is exclusively for one component it's not necessary to add it in the E2E tests repo

As an example of what I meant, https://github.com/kubewarden/kubewarden-controller/pull/139/files is a feat PR that contains `e2e-tests/test-verification/`. This made it simple to explain what was expected from the feature, and to iterate on the implementation 3 times on the PR.

> For the component repo, as long as the developer abstracts the two different test frameworks under a helper script or Makefile target, I do not see that as a problem. 

That sounds like the interface I want specified: "a repo must have a `make e2e-tests` makefile target that succeeds with retcode 0 if all tests pass". And then the debugger can run those for each repo, in a specific order. We can then write a simple linter to make sure we never deviate from the interface, and break `git bisect`, GH Actions, etc in the future.

> For the E2E test, it uses versions available somewhere. So, if you are running a bisect, you can publish the version under the bisect somewhere (probably a registry)

I would like to define that place for the interface. Is it a config file in the e2e-tests repo, a registry, does it accept both, a mix?


As a side note, I think that we are seeing 2 approaches, you would like the "interface" to be defined organically by the implementation of the e2e tests, and I would like to specify it beforehand.

Reply:
What do you think about start by trying the following idea and iterating over it? 

- For PRs we follow the @ereslibre suggestion and run the e2e test using some kind of GH action (probable defined in the `kubewarden-end-to-end-tests` repository) 
 - For local debugging, we do something similar of it's done `kubewarden-controller`. In the `mate e2e-test` script we download the e2e tests locally (no sub-modules for now) and run them using the same interface used by the GH action. Just to clarify, the `kubewarden-controller` does not download tests from another repository, but it [download a script to setup the test env locally](https://github.com/kubewarden/kubewarden-controller/blob/main/Makefile#L59-L62) from another repository (provided by controller-runtime).

Reply:
@viccuad I've updated the RFC adding an idea of how to run the e2e tests from other repositories defining some interface to allow this integration. Can you review it?